modin/core/dataframe/pandas/dataframe/utils.py

```diff
@@ -0,0 +1,266 @@
+
+    We use this method to provide backwards compatibility with NumPy versions < 1.23 (e.g. when
+    the user is using Modin in compat mode). This is basically a wrapper around `np.quantile` that
+    ensures we provide the correct `method` argument - i.e. if we are dealing with objects (which
+    may or may not support algebra), we do not want to use a method to find quantiles that will
+    involve algebra operations (e.g. mean) between the objects, since that may fail.
+
+    Parameters
+    ----------
+    df : pandas.DataFrame or pandas.Series
+        The data to pick quantiles from.
+    quantiles : list[float]
+        The quantiles to compute.
+    method : str
+        The method to use. `linear` if dealing with numeric types, otherwise `inverted_cdf`.
+
+    Returns
+    -------
+    np.ndarray
+        A NumPy array with the quantiles of the data.
+    """
+    if method == "linear":
+        # This is the default method for finding quantiles, so it does not need to be specified,
+        # which keeps backwards compatibility with older versions of NumPy that do not have a
+        # `method` keyword argument in np.quantile.
+        return np.quantile(df, quantiles)
+    else:
+        try:
+            return np.quantile(df, quantiles, method=method)
+        except Exception:
+            # In this case, we're dealing with an array of objects, but the current version of
+            # NumPy does not have a `method` kwarg. We need to use the older kwarg, `interpolation`
+            # instead.
+            return np.quantile(df, quantiles, interpolation="lower")
+
+
+def pick_samples_for_quantiles(
+    df: pandas.DataFrame,
+    column: str,
+    num_partitions: int,
+    length: int,
+) -> np.ndarray:
+    """
+    Pick samples over the given partition.
+
+    This function picks samples from the given partition using the TeraSort algorithm - each
+    value is sampled with probability 1 / m * ln(n * t) where m = total_length / num_partitions,
+    t = num_partitions, and n = total_length.
+
+    Parameters
+    ----------
+    df : pandas.Dataframe
+        The masked dataframe to pick samples from.
+    column : str
+        The columns to pick quantiles from. Only the first column in the list will be used.
+    num_partitions : int
+        The number of partitions.
+    length : int
+        The total length.
+
+    Returns
+    -------
+    np.ndarray:
+        The samples for the partition.
+
+    Notes
+    -----
+    This sampling algorithm is inspired by TeraSort. You can find more information about TeraSort
+    and the sampling algorithm at https://www.cse.cuhk.edu.hk/~taoyf/paper/sigmod13-mr.pdf.
+    """
+    m = length / num_partitions
+    probability = (1 / m) * np.log(num_partitions * length)
+    return df[column].sample(frac=probability).to_numpy()
+
+
+def pick_pivots_from_samples_for_sort(
+    df: PandasDataframe,
+    samples: np.ndarray,
```

Comment:
are we sure `samples` are an array and not a list of arrays?

Reply:
Good point - I'll fix it!